tsdate/util.py

```diff
@@ -344,9 +370,48 @@ def mutation_span_array(tree_sequence):
     return mutation_spans, mutation_edges
 
 
-@numba.njit(_unituple(_i1w, 3)(_i1r, _i1r, _f1r, _f1r, _b1r))
+# Some functions for changing tskit metadata
+# See https://github.com/tskit-dev/tskit/discussions/2954
+def _change_ragged_order(ragged_arr, ragged_offset, new_order):
+    # returns a tuple of a new ragged_col in a new order and a new offset array
+    old_ranges = np.array([ragged_offset[:-1], ragged_offset[1:]]).T
+    ranges = old_ranges[new_order, :]
+    ix = [np.arange(lf, rt, dtype=ragged_offset.dtype) for lf, rt in ranges if lf != rt]
```

Comment:
Because of this comprehension, this function seems like a good candidate for numba-fication. Have you benchmarked this routine on very large treeseqs, versus the tskit.unpack_strings approach?

Reply:
Good point. I guess there are 2 scenarios

1. large ts with no node metadata (easy e.g. to simulate)
2. large ts with metadata (perhaps one of the unified genealogy files, e.g. https://zenodo.org/records/5512994/files/hgdp_tgp_sgdp_high_cov_ancients_chr2_q.dated.trees.tsz)

Reply:
```python
from datetime import datetime
import tszip
ts = tszip.decompress("hgdp_tgp_sgdp_high_cov_ancients_chr2_q.dated.trees.tsz")

def _change_ragged_order(ragged_arr, ragged_offset, new_order):
    # returns a tuple of a new ragged_col in a new order and a new offset array
    old_ranges = np.array([ragged_offset[:-1], ragged_offset[1:]]).T
    ranges = old_ranges[new_order, :]
    ix = [np.arange(lf, rt, dtype=ragged_offset.dtype) for lf, rt in ranges if lf != rt]
    select = [] if len(ix) == 0 else np.concatenate(ix)
    return ragged_arr[select], np.insert(np.cumsum(np.diff(ranges, axis=1)), 0, 0)

def _change_ragged_order2(ragged_arr, ragged_offset, new_order):
    # returns a tuple of a new ragged_col in a new order and a new offset array
    arr = tskit.unpack_arrays(ragged_arr, ragged_offset)
    return tskit.pack_arrays([arr[i] for i in new_order])

def _change_ragged_order3(ragged_arr, ragged_offset, new_order):
    # returns a tuple of a new ragged_col in a new order and a new offset array
    arr = tskit.unpack_bytes(ragged_arr, ragged_offset)
    return tskit.pack_bytes([arr[i] for i in new_order])


nodes = ts.tables.nodes
start = datetime.now()
a = _change_ragged_order3(nodes.metadata, nodes.metadata_offset, ts.num_nodes - np.arange(ts.num_nodes) - 1)
print(f"took {datetime.now() - start} secs")
```

Reply:
Actually, if I try this with a large ts with no node metadata (e.g. `ts = msprime.simulate(10000, recombination_rate=10000)`), then `_change_ragged_order2` takes about 4 times longer than my `_change_ragged_order` original version. So 🤷

Reply:
Actually, this would seem to hit the common use-cases:

```python
def change_ragged_order(ragged_arr, ragged_offset, new_order):
    # returns a tuple of a new ragged_col in a new order and a new offset array
    if len(ragged_arr) == 0:
        # Common edge case: no metadata => offset array is just zeros
        return ragged_arr, np.zeros(len(new_order), dtype=ragged_offset.dtype)
    tskit.unpack_arrays(ragged_arr, ragged_offset)
    return tskit.pack_arrays([arr[i] for i in new_order])

```

Reply:
Changed now.